command/src/certificate.rs

```diff
@@ -1,18 +1,69 @@
-use std::{error, fmt, str::FromStr};
+use std::{collections::HashSet, error, fmt, str::FromStr};
 
 use hex::FromHex;
-use pem::parse;
 use serde::de::{self, Visitor};
 use sha2::{Digest, Sha256};
+use x509_parser::{
+    oid_registry::{OID_X509_COMMON_NAME, OID_X509_EXT_SUBJECT_ALT_NAME},
+    pem::{parse_x509_pem, Pem},
+};
 
 use crate::proto::command::TlsVersion;
 
-#[derive(thiserror::Error, Debug)]
+// -----------------------------------------------------------------------------
+// CertificateError
+
+#[derive(thiserror::Error, Clone, Debug)]
 pub enum CertificateError {
     #[error("Could not parse PEM certificate from bytes: {0}")]
     InvalidCertificate(String),
+    #[error("failed to parse certificate common name attribute, {0}")]
+    InvalidCommonName(String),
+    #[error("failed to parse certificate subject alternate names attribute, {0}")]
+    InvalidSubjectAlternateNames(String),
+}
+
+// -----------------------------------------------------------------------------
+// parse
+
+pub fn parse(certificate: &[u8]) -> Result<Pem, CertificateError> {
+    let (_, pem) = parse_x509_pem(certificate)
+        .map_err(|err| CertificateError::InvalidCertificate(err.to_string()))?;
+
+    Ok(pem)
+}
+
+// -----------------------------------------------------------------------------
+// get_cn_and_san_attributes
+
+pub fn get_cn_and_san_attributes(pem: &Pem) -> Result<HashSet<String>, CertificateError> {
+    let x509 = pem
+        .parse_x509()
+        .map_err(|err| CertificateError::InvalidCertificate(err.to_string()))?;
+
+    let mut names: HashSet<String> = HashSet::new();
+    for name in x509.subject().iter_by_oid(&OID_X509_COMMON_NAME) {
+        names.insert(
+            name.as_str()
+                .map(String::from)
+                .map_err(|err| CertificateError::InvalidCommonName(err.to_string()))?,
+        );
+    }
+
+    for name in x509.subject().iter_by_oid(&OID_X509_EXT_SUBJECT_ALT_NAME) {
+        names.insert(
+            name.as_str()
+                .map(String::from)
+                .map_err(|err| CertificateError::InvalidSubjectAlternateNames(err.to_string()))?,
+        );
+    }
+
+    Ok(names)
 }
 
+// -----------------------------------------------------------------------------
+// ParseErrorTlsVersion
```

Comment:
these lines and comments are useless.

instead of a `// ParseErrorTlsVersion` over a `ParseErrorTlsVersion` struct, why not provide a documenting comment with `///` instead, that will explain what this type is?